src/plugins/data_views/public/data_views/data_views_api_client.ts

```diff
@@ -6,33 +6,61 @@
  * Side Public License, v 1.
  */
 
-import { HttpSetup } from '@kbn/core/public';
+import { HttpSetup, HttpResponse } from '@kbn/core/public';
 import { DataViewMissingIndices } from '../../common/lib';
 import { GetFieldsOptions, IDataViewsApiClient } from '../../common';
 import { FieldsForWildcardResponse } from '../../common/types';
-import { FIELDS_FOR_WILDCARD_PATH } from '../../common/constants';
+import { FIELDS_FOR_WILDCARD_PATH, FIELDS_PATH } from '../../common/constants';
 
 const API_BASE_URL: string = `/api/index_patterns/`;
 const version = '1';
 
+async function sha1(str: string) {
+  const enc = new TextEncoder();
+  const hash = await crypto.subtle.digest('SHA-1', enc.encode(str));
+  return Array.from(new Uint8Array(hash))
+    .map((v) => v.toString(16).padStart(2, '0'))
+    .join('');
+}
+
 /**
  * Data Views API Client - client implementation
  */
 export class DataViewsApiClient implements IDataViewsApiClient {
   private http: HttpSetup;
+  private getCurrentUserId: () => Promise<string | undefined>;
 
   /**
    * constructor
    * @param http http dependency
    */
-  constructor(http: HttpSetup) {
+  constructor(http: HttpSetup, getCurrentUserId: () => Promise<string | undefined>) {
     this.http = http;
+    this.getCurrentUserId = getCurrentUserId;
   }
 
-  private _request<T = unknown>(url: string, query?: {}, body?: string): Promise<T | undefined> {
+  private async _request<T = unknown>(
+    url: string,
+    query?: {},
+    body?: string,
+    forceRefresh?: boolean
+  ): Promise<HttpResponse<T> | undefined> {
+    const asResponse = true;
+    const cacheOptions = forceRefresh ? { cache: 'no-cache' as RequestCache } : {};
+    const userId = await this.getCurrentUserId();
+
+    const userHash = userId ? await sha1(userId) : '';
+
     const request = body
-      ? this.http.post<T>(url, { query, body, version })
-      : this.http.fetch<T>(url, { query, version });
+      ? this.http.post<T>(url, { query, body, version, asResponse })
+      : this.http.fetch<T>(url, {
+          query,
+          version,
```

Comment:
ES response changes wouldn't change anything but Kibana response changes would. If we passed `hasData` through to the kibana response, then the etag would change, resulting in a 200 instead of a 304 response. That said, we'd still need to wait for the cache to go stale before the request will be fired.